contracts/modules/v2/StakingPool.sol

```diff
@@ -173,145 +174,238 @@ contract StakingPool is IStakingPool, ERC721 {
     }
 
-    /* update rewards */
+    position.lastAccNxmPerRewardShare = _accNxmPerRewardsShare;
 
-    uint _rewardPerSecond = rewardPerSecond;
-    uint elapsed = block.timestamp - lastRewardUpdate;
+    /* end calculate and update pending reward */
 
-    if (elapsed > 0) {
-      lastRewardUpdate = block.timestamp;
-      accRewardPerPoolShare += elapsed * _rewardPerSecond / _rewardsSharesSupply;
-    }
+    uint newStakeShares = _stakeSharesSupply == 0
+      ? sqrt(amount)
+      : _stakeSharesSupply * amount / _activeStake;
 
-    /* store */
+    uint newRewardsShares = calculateRewardSharesAmount(newStakeShares, groupId);
 
     group.stakeShares += newStakeShares;
     group.rewardsShares += newRewardsShares;
-    group.groupSharesSupply += newGroupShares;
 
-    stakeGroups[groupId] = group;
-    balanceOf[positionId][groupId] += newGroupShares;
+    /* storage writes */
 
+    // update globals
     activeStake = _activeStake + amount;
     stakeSharesSupply = _stakeSharesSupply + newStakeShares;
     rewardsSharesSupply = _rewardsSharesSupply + newRewardsShares;
+
+    // update group and staker data
+    groups[groupId] = group;
+    positions[tokenId][groupId] = position;
+    pendingRewards[tokenId] = pendingReward;
+
+    /* end storage writes */
   }
 
-  function withdraw(WithdrawParams[] memory params) external {
-    // 1. check nft ownership / allowance
-    // 2. loop through each group:
-    // 2.1. check group expiration if the deposit flag is set
-    // 2.2. calculate the reward amount if the reward flag is set
-    // 3. sum up nxm amount of each group
-    // 4. transfer nxm to staker
+  function calculateRewardSharesAmount(
+    uint stakeSharesAmount,
+    uint groupId
+  ) internal view returns (uint) {
+
+    uint lockDuration = (groupId + 1) * GROUP_SIZE - block.timestamp;
+    uint maxLockDuration = GROUP_SIZE * 8;
+
+    // TODO: determine extra rewards formula
+    return
+      stakeSharesAmount
+      * REWARDS_SHARES_RATIO
+      * lockDuration
+      / REWARDS_SHARES_DENOMINATOR
+      / maxLockDuration;
+  }
+
+  function withdraw(WithdrawParams[] calldata params) external {
+
+    updateGroups();
+
+    uint _accNxmPerRewardsShare = accNxmPerRewardsShare;
+    uint _firstActiveGroupId = block.timestamp / GROUP_SIZE;
+
+    for (uint i = 0; i < params.length; i++) {
+
+      uint stakeToWithdraw;
+      uint rewardToWithdraw;
+
+      uint tokenId = params[i].tokenId;
+      uint groupCount = params[i].groupIds.length;
+
+      for (uint j = 0; j < groupCount; j++) {
+
+        uint groupId = params[i].groupIds[j];
+        Position memory position = positions[tokenId][groupId];
+
+        // can withdraw stake only if the group is expired
+        if (params[i].withdrawStake && groupId < _firstActiveGroupId) {
+
+          // calculate the amount of nxm for this position
+          uint stake = expiredGroups[groupId].stakeAmountAtExpiry;
```

Comment:
The stake is always in NXM. If it's shares it will contain "shares" in its name.

Reply:
The point was to consistently use the denomination at the end of the variable in a similar fashion to the other contracts where we suffix them with -InDemonination to dispel any doubts. 

While you're stating that the stake is always in NXM, the code itself does not make that very obvious. Since it did make me doubt the denomination while I was reading it, it is as likely to happen to anyone else who is reading the code.

The purpose of code reviews is often times to adapt the code in such a way that everyone else can easily grasp it. While everything might make sense it does not mean it's as obvious for anyone else.

Reply:
I think the denomination makes sense when we have multiple assets, but it looks clear enough to me in this context.